### PR TITLE
feat: Add OpenTelemetry trace export support for torch.profiler

### LINKED ISSUE #153614: Add opentelemetry traces
### 🚀 The feature, motivation and pitch

Hello,

`opentelemetry-cpp` is a dependency of torch, but I see no trace (pun intended) of it in the code base.

We are adding `opentelemetry` to [transformers](https://github.com/huggingface/transformers) in order to profile a new feature that is performance sensitive, and we might extended further later on if it proves to be useful.

When profiling code that uses pytorch with OTel, we lack details of what specific operations were talking the most time. My current approach was to extract the profiling data from pytorch using the Kineto profiler, but traces will not be  attached to the correct parent span messing up the readability of a given trace. Additionally, given the number of profiling events, converting each to a span ends up making a large number of spans (IIRC, a trace shouldn't go over 50mb in size for efficiency).

One solution opted by frameworks using OTel is to provide an Instrumentation package that exposes an instrumentor class, `{LibraryName}Instrumentor`, so that users have detailed traces on the inner working of the library (cf [OpenTelemetry-Python-Contrib](https://opentelemetry-python-contrib.readthedocs.io/en/latest/index.html) or [openinference-instrumentation-mistralai](https://github.com/Arize-ai/openinference/tree/main/python/instrumentation/openinference-instrumentation-mistralai)).

Not sure an external package for the instrumentor class will work, given the core of pytorch is in c++ (I think external packages work thanks to monkey patching in Python).

Happy to hear more feedback from others on this!

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Nik-Reddy**
## Summary

Adds opentelemetry_trace_handler() to torch.profiler that converts PyTorch profiler events into OpenTelemetry spans, enabling seamless integration with any OTel-compatible observability backend (Jaeger, Zipkin, OTLP, Datadog, etc.).

Fixes #153614

## Motivation

As discussed in #153614, PyTorch's profiler produces rich tracing data via Kineto, but there's no way to export it as OpenTelemetry spans. Users profiling PyTorch workloads with OTel lack visibility into which specific operations are taking the most time, and manually converting Kineto traces results in spans detached from the correct parent context.

This PR bridges that gap by providing a first-class `on_trace_ready` handler that follows the same pattern as `tensorboard_trace_handler`.

### API (`torch/profiler/_opentelemetry.py`)

```python
from torch.profiler import profile, schedule, opentelemetry_trace_handler

handler = opentelemetry_trace_handler(
    service_name="my-training-job",       # OTel service name
    tracer_provider=my_provider,          # optional: bring your own TracerProvider
    span_exporter=OTLPSpanExporter(),     # optional: any SpanExporter
)

with profile(
    activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
    schedule=schedule(wait=1, warmup=1, active=2),
    on_trace_ready=handler,
    record_shapes=True,
    with_stack=True,
) as prof:
    for step in range(N):
        train_step()
        prof.step()
```

### Key decisions

| Decision | Rationale |
|----------|-----------|
| **Optional dependency** | `opentelemetry-api` / `opentelemetry-sdk` are imported at call time with a clear `ImportError` if missing |
| **`on_trace_ready` callback** | Follows the established `tensorboard_trace_handler` pattern — zero changes to the profiler core |
| **Parent–child spans** | Walks the profiler's existing `cpu_parent`/`cpu_children` tree to create properly nested OTel spans |
| **Rich attributes** | Each span carries `pytorch.*` attributes: timing, device type/index, input shapes, dtypes, FLOPs, memory usage, stack traces, kernel names |
| **Pluggable export** | Users can inject any `TracerProvider` or `SpanExporter` — works with OTLP, Jaeger, Zipkin, console, or custom backends |

### Span attributes exported

- `pytorch.event.id`, `pytorch.thread.id`, `pytorch.scope`
- `pytorch.device.type`, `pytorch.device.index`, `pytorch.device.name`
- `pytorch.cpu_time_total_us`, `pytorch.self_cpu_time_total_us`
- `pytorch.device_time_total_us`, `pytorch.self_device_time_total_us`
- `pytorch.cpu_memory_usage_bytes`, `pytorch.device_memory_usage_bytes`
- `pytorch.flops`, `pytorch.input_shapes`, `pytorch.input_dtypes`
- `pytorch.stack`, `pytorch.kernel_names`
- `pytorch.is_async`, `pytorch.is_user_annotation`

## Files changed

- **`torch/profiler/_opentelemetry.py`** - New module with the exporter implementation
- **`torch/profiler/__init__.py`** - Export `opentelemetry_trace_handler` in public API
- **`test/profiler/test_opentelemetry.py`** - 14 tests covering unit helpers, attribute extraction, end-to-end export with in-memory exporter, parent-child relationships, timing validation, and graceful handling of missing OTel

## Test plan

All 14 tests pass:
`
$ python -m pytest test/profiler/test_opentelemetry.py -v
14 passed in 0.45s
`

Tests cover:
- Import error when OTel is missing
- Helper functions (`_us_to_ns`, `_safe_str`, `_SCOPE_NAMES`)
- Attribute extraction from real `FunctionEvent`
- Event tree construction (root identification)
- End-to-end handler with in-memory span exporter
- Explicit `TracerProvider` injection
- Parent-child span relationships
- Empty events (no crash)
- Span timing validation
- Public API accessibility (import + `__all__`)

cc @robieta @chaekit @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise @mwootton @divyanshk @jiannanWang @scotts @ryanzhang22

### CODE REVIEW

**Comment by Nik-Reddy** [comment]
Any chance someone could take a look at this? 
 
 It lets callers skip the input validation in `torch.multinomial` via  `validate=False`, which matters in LLM inference where the input comes straight from softmax and validation eats ~66% of the GPU time on every decode step. Defaults to True so nothing changes for existing code.

**Comment by Nik-Reddy** [comment]
Hi @scotts, this has been open since mid-April and has not had a first pass yet. CI is green and the change is opt-in (off by default, no behavior change unless the exporter is configured). Would you have bandwidth to take a look, or is there a better owner in the profiler area I should tag instead?
